src/commands/mod.rs

```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+                .iter()
+                .map(|txin| txin.previous_output)
+                .collect();
+            let prev_coins = db_conn.coins_by_outpoints(prev_outpoints);
+            assert_eq!(prev_outpoints.len(), prev_coins.len());
+            prev_coins
+        };
+        if prev_coins
+            .iter()
+            .any(|(_, coin)| coin.spend_block.is_some())
+        {
+            return Err(CommandError::RbfError(RbfErrorInfo::TxAlreadyConfirmed(
+                *txid,
+            )));
+        }
+        let mempool_entry = self
+            .bitcoin
+            .mempool_entry(txid)
+            .ok_or(CommandError::RbfError(RbfErrorInfo::TxNotYetBroadcast(
+                *txid,
+            )))?;
+        // Check sum of descendant counts of all directly conflicting transactions is not too high.
+        // We are assuming we only have one directly conflicting transaction.
+        if mempool_entry.descendant_count > RBF_MAX_DESCENDANT_COUNT {
+            return Err(CommandError::RbfError(RbfErrorInfo::TooManyDescendants(
+                mempool_entry.descendant_count,
+            )));
+        }
+        let prev_feerate_vb = {
+            let prev_abs_fee = prev_psbt
+                .fee()
+                .expect("fee must be Ok as transaction is already in mempool");
+            let prev_tx_size: u64 = prev_psbt
+                .clone()
+                .extract_tx()
+                .vsize()
+                .try_into()
+                .expect("vsize must fit in a u64");
+            prev_abs_fee
+                .checked_div(prev_tx_size)
+                .ok_or(CommandError::InsaneFees(InsaneFeeInfo::InvalidFeerate))?
+        };
+        // Check replacement transaction's feerate is greater than previous feerate (the only directly conflicting transaction).
+        // TODO: Should we throw error or just set feerate_vb = prev_feerate_vb + 1 and check `rbf_psbt` below?
+        if feerate_vb < prev_feerate_vb.to_sat().checked_add(1).unwrap() {
+            return Err(CommandError::RbfError(RbfErrorInfo::TooLowFeerate(
+                feerate_vb,
+            )));
+        }
+        let prev_txos = &prev_psbt.unsigned_tx.output;
+        let mut destinations: HashMap<bitcoin::Address, bitcoin::Amount> =
+            HashMap::with_capacity(prev_txos.len());
+        let mut change_index: Option<bip32::ChildNumber> = None;
+        for txo in prev_txos {
+            let address = bitcoin::Address::from_script(
+                &txo.script_pubkey,
+                self.config.bitcoin_config.network,
+            )
+            .map_err(CommandError::Address)?;
+            match db_conn.derivation_index_by_address(&address) {
+                Some((ind, true)) => {
+                    // TODO: check if there is more than one change output?
+                    println!("found change at address {} and index {}", address, ind);
+                    change_index = Some(ind);
+                }
```

Comment:
My general assumption has been that change addresses are added automatically to the transaction as change output and not manually used as destinations by the user, but that's not guaranteed, even if a single change address is used. Perhaps in case there's more than one change address used, we could preserve all of them and not treat any as change output.

Reply:
> As discussed, if more than one of our change addresses is used in the existing transaction, we can keep one of them to use for change in the replacement transaction (and drop the others?).

No i don't think we should drop the others. We should just keep them like any other destinations.

> Perhaps the choice of change address to use in the replacement should not depend on its value, but rather its index? E.g. choosing the highest index so that there's less chance of reusing change addresses in a future transaction.

That's a good point but i think it's still better to choose the largest value change output, to take the fees from there to avoid having to increase the transaction size.

----

In general my thinking is: if the user has set the outputs the way they are, he has a reason. Let's avoid touching the structure of the outputs *except* if we can find a change output suitable to be reduced in order to pay for the fees, to avoid including more inputs and/or an additional change output.

Reply:
I changed the logic so that it chooses the change address with the highest value and then the one with the highest index, in case there's more than one. The other change addresses are then treated in the same way as non-change outputs.